### PR TITLE
Check block in mutable storage

### PR DESCRIPTION
**Pull request by kamilsa**
### Description of the Change

In chain validator we might need to check some logic on either EmptyBlock or Block types. Current interface of mutable storage does not allow that. To overcome this issue check method has been introduced which applies predicates logic on the block without actually applying the block to the storage.

### Benefits

We can validate block on storage without applying it

### Possible Drawbacks 

Probably we don't need check method to take AbstractBlock as parameter but use EmptyBlock instead.

### Usage Examples or Tests

Mutable storage test has been introduced

### CODE REVIEW

File: test/module/irohad/ametsuchi/mutable_storage_test.cpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                [&expected_block, &expected_res](
+                    const auto &block, const auto &, const auto &) {
+                  EXPECT_EQ(expected_block, block);
+                  return expected_res;
+                }));
+}
+
+INSTANTIATE_TEST_CASE_P(MutableStorageParameterizedTest,
+                        MutableStorageTest,
+                        Bool(), );
```
Comment (kamilsa) [inline test/module/irohad/ametsuchi/mutable_storage_test.cpp]: comma is needed to make it compile https://github.com/google/googletest/issues/1419

**Comment by l4l** [comment]
please write meaningful pr title

File: irohad/ametsuchi/mutable_storage.hpp [hunk 2]
```diff
@@ -38,6 +39,22 @@ namespace iroha {
      */
     class MutableStorage {
      public:
+      template <typename T>
+      using MutableStoragePredicateType =
```
Comment (muratovv) [inline irohad/ametsuchi/mutable_storage.hpp]: Add documentation for using.

File: irohad/ametsuchi/mutable_storage.hpp [hunk 3]
```diff
@@ -38,6 +39,22 @@ namespace iroha {
...
+                             const shared_model::interface::types::HashType &)>;
+
+      /**
+       * Checks if block satisfies predicated
+       * @param block block to be checked
+       * @param predicate function returning true if predicate satisfied and
+       * false otherwise
+       * @return result of predicate
+       */
+      virtual bool check(const shared_model::interface::AbstractBlock &block,
```
Comment (muratovv) [inline irohad/ametsuchi/mutable_storage.hpp]: Maybe rework abstract block with block variant?

File: test/module/irohad/ametsuchi/mutable_storage_test.cpp (hunk 1 above)
Comment (muratovv) [reply test/module/irohad/ametsuchi/mutable_storage_test.cpp]: Maybe put a comment in the code?

Comment (muratovv) [reply test/module/irohad/ametsuchi/mutable_storage_test.cpp]: Maybe put a comment in the code?

**Pull request merged**